File: Combination-Sum-II/CS2.clj
(defn combination-sum2 [cands target]
  (let [cands (sort cands)
        list-of-sets (doall (for [_ (range (inc target))]
                              (atom #{})))]
    
    (swap! (nth list-of-sets 0) (fn [x] (conj x [])))

    (doseq [cand cands
            t (range target (dec cand) -1)
            prev @(nth list-of-sets (- t cand))]

      (swap! (nth list-of-sets t)
             conj
             (conj prev cand)))
    
    @(last list-of-sets)
    ))

File: Combination-Sum-II/CS2.jl
function combinationSum2(cands, target)
    sort!(cands) # sort all cands
    listOfSets = [Set() for _ in 1:(target+1)]

    push!(listOfSets[1], []) # index 1 actually point to 0, because julia index start from 1

    for
        cand in cands,
        t in range(target, cand, step=-1),
        prev in listOfSets[t+1-cand];

        cache = copy(prev)
        push!(cache, cand)
        push!(listOfSets[t+1], cache)
    end
        
    last(listOfSets)
end
